plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/S3BlobStoreContainerTests.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cloud.aws.blobstore;
+
+import com.amazonaws.Protocol;
+import com.amazonaws.services.s3.AmazonS3;
+import org.elasticsearch.cloud.aws.InternalAwsS3Service;
+import org.elasticsearch.common.blobstore.BlobStore;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.repositories.ESBlobStoreContainerTestCase;
+
+import java.io.IOException;
+import java.util.Locale;
+
+public class S3BlobStoreContainerTests extends ESBlobStoreContainerTestCase {
+    protected BlobStore newBlobStore() throws IOException {
+        int maxRetries = 5;
+        String region = null;
+        Settings settings = Settings.EMPTY;
+
+        InternalAwsS3Service service = new InternalAwsS3Service(settings);
+        AmazonS3 client = service.client(null, Protocol.HTTPS, region, null, null, maxRetries);
+        String bucket = randomAsciiOfLength(randomIntBetween(1, 10)).toLowerCase(Locale.ROOT);
+
+        return new S3BlobStore(settings, client, bucket, region, false,
+            new ByteSizeValue(100), maxRetries, "public-read-write", "standard");
+    }
```

Comment:
Looking further, we actually have an AmazonS3 test class already, look
at: org.elasticsearch.cloud.aws.TestAmazonS3.  Its used in
TestAwsS3Service, which as far as I can tell, is unused.  TestAmazonS3
looks to provide what you would need for creating a real test against S3,
as @imotov mentioned.  AzureBlobStoreTests already has this implemented,
which could be used for a container class test.  Looking at the Google mock
a bit more closely, it seems to be testing against the google APIs
themselves.

If we are testing against real repositories from time to time, it probably
takes away the need to mock the blob store implementations themselves.  In
that case, we should just implement a test against the real back-end
storage mechanism, and run it before finalizing this PR.  It seems like the
infrastructure for this is already in place, though I'm not sure where the
TestAwsS3Service/TestAmazonS3 stuff is used.

I would wait for @dadoonet to provide details before spending much more
time on mocking things that may not be relevant.

Sorry for the changes; thank you for your effort on this!



On Sun, Jun 12, 2016 at 6:51 PM, gfyoung <notifications@github.com> wrote:

> In
> plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/blobstore/S3BlobStoreContainerTests.java
> <https://github.com/elastic/elasticsearch/pull/18815#discussion_r66729195>
> :
>
> > +import java.io.IOException;
> > +import java.util.Locale;
> > +
> > +public class S3BlobStoreContainerTests extends ESBlobStoreContainerTestCase {
> > +    protected BlobStore newBlobStore() throws IOException {
> > +        int maxRetries = 5;
> > +        String region = null;
> > +        Settings settings = Settings.EMPTY;
> > +
> > +        InternalAwsS3Service service = new InternalAwsS3Service(settings);
> > +        AmazonS3 client = service.client(null, Protocol.HTTPS, region, null, null, maxRetries);
> > +        String bucket = randomAsciiOfLength(randomIntBetween(1, 10)).toLowerCase(Locale.ROOT);
> > +
> > +        return new S3BlobStore(settings, client, bucket, region, false,
> > +            new ByteSizeValue(100), maxRetries, "public-read-write", "standard");
> > +    }
>
> @imotov <https://github.com/imotov> : That would be great to have, since
> that would facilitate the test setup for those repositories. However, I
> would still keep mocking on the table as @abeyad
> <https://github.com/abeyad> proposed, since that's what the Google cloud
> tests do with MockHttpTransport.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/18815/files/5ecfabc0dffffcbec1cbf7c5c7db0e3ab05a8c07#r66729195>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABjkQUhvZEbCRkInDGZvlL7FzEEQ3S6Fks5qLI1vgaJpZM4Iy-tw>
> .
>